### ISSUE TITLE
OIDC - Authenticate with token in Cookie or Url Parameter 

### ISSUE THREAD

**Issue by DorianMaliszewski** [quarkusio/quarkus#7414]
**Description**
OIDC work with Authorization header, it can be nice to authenticate users through Url parameter or Cookie, to implements easily Websocket security or SSE security.

**Implementation ideas**
Before any process to validate the token check if it is in Authorization header / Url Parameter (access_token for example) / Cookie name set in application.properties.

This is already implemented in smallrye-jwt but not in the oidc and when we use the two security it always shows 401 or 403 error, so...

**Thank you in advance**
_I'm sorry if someone already requested this feature, I don't see it in the issues_

**Comment by sberyozkin** [comment]
Hi @DorianMaliszewski I'm not sure about it because `quarkus-oidc` is about supporting the standard OIDC flows with `Authorization` being a standard carrier.
I'm assuming a token which might be available in a Cookie (or URI parameter) has not been issued by the OIDC server ?

May be instead we should come up with a way for the `quarkus-oidc` and `quarkus-smallrye-jwt` co-exist, if the former fails, the the next one is tried. Such that multiple `HttpAuthenticationMechanism`s can be checked ? Not 100% sure how messy it can become though :-)
CC @stuartwdouglas @pedroigor

**Comment by DorianMaliszewski** [comment]
@sberyozkin Hi,
Yeah I don't know what can be the best solution. But today I'm facing this issue on Quarkus, I can't pass the access_token from Websocket or EventSource to authenticate the user on Quarkus OIDC 🤔

May be there are a better solution ? 😅

**Comment by sberyozkin** [comment]
@DorianMaliszewski But what is the origin of the token ? `quarkus-oidc` is for the tokens issued by the OIDC providers like Keycloak.

> May be there are a better solution ?

I'm not sure :-) but as I said, perhaps having a queue of the authentication mechanisms might work

**Comment by sberyozkin** [comment]
@DorianMaliszewski But what is the origin of the token ? `quarkus-oidc` is for the tokens issued by the OIDC providers like Keycloak.

> May be there are a better solution ?

I'm not sure :-) but as I said, perhaps having a queue of the authentication mechanisms might work

**Comment by sberyozkin** [comment]
May be emulate `Authorization` header when it is not on the HTTP upgrade path (web-socket), etc ?

**Comment by DorianMaliszewski** [comment]
I'm authenticating on my API with an access token issued by Keycloak.
I put this token in all my HTTP request in the Authorization header.
But with Websocket and EventSource it's not possible to put some headers 😭. (According to MDN documentation)
So I always get 403 or 401 response when I want to connect with WS or ES @sberyozkin.

If it's possible, please it will be greater for everyone like to have a tips or exemple 😅

**Comment by DorianMaliszewski** [comment]
I use quarkus-oidc and quarkus-undertow (or something like that)

**Comment by sberyozkin** [comment]
Hi @DorianMaliszewski Thanks for the update. I'm really not sure it makes sense to have `quarkus-oidc` covering the web sockets and, once the upgrade has been done it is no longer HTTP, and as such some other mechanism should be used.
Lets see what Stuart and Pedro think

**Comment by pedroigor** [comment]
Accordingly to https://tools.ietf.org/html/rfc6750#section-2.3, query parameters should be avoided.

Cookies are vulnerable to CSRF as you may have iframes being used in other domains and using your endpoint like if it were part of your domain.

I'm not sure how to properly handle this too. Some time ago I tried something where you would hack `BASIC` authentication to include only the token instead of username/password. IIRC, one of the issues was that after authenticating during the handshake, I could not check anymore whether or not the token was still valid.

What if we provide some hook for people so that they can resolve the token from wherever they want? For some people, use a query parameter should be fine but we would still don't recommend it. Or maybe dedicate one sprint to try to come up with something we think is better.

**Comment by sberyozkin** [comment]
Hi Pedro, thanks for the feedback. I've been thinking it is out of scope for the OIDC itself because it is not really OIDC any longer :-), we can't even enforce at the OIDC level it is the secure web sockets channel (should probably have an option to enforce it is HTTPS in `quarkus-oidc` :-) ).

Hi @DorianMaliszewski What is the actual flow here, the user authenticates to Quarkus 'service' application first, and then in some follow up request a web socket upgrade is done ? is the upgrade request accompanied by the bearer token ? If yes then may be we can have that token stored in the corresponding binding on the server side and reuse it every time, but I have to admit I'm out of depth here :-) and would be glad to hear from Stuart @stuartwdouglas as well

**Comment by DorianMaliszewski** [comment]
@sberyozkin @pedroigor 
Hi,

In my project I use Keycloak, to get an access_token on my frontend, (like React or Vue) then when I have it, I want to make a Websocket connection or Event Source connection on my API, which use `quarkus-oidc` and `quarkus-undertow`. On this API there are Http Endpoint (on which I can authenticate through the Authorization header resolved by `quarkus-oid`) but for SSE and Websocket endpoint, I can't authenticate my user securely. :disappointed: 

I have a suggestion let me know what do you think about it :

- Make a filter before token validation on incoming request and check if it's a ws/wss protocol or produces "text/event-stream"
- If yes, check token parameter in url and put in Authorization header to make `quarkus-oidc` check it
- If no, do nothing

What do you think ? :sweat_smile:

**Comment by sberyozkin** [comment]
Hi @DorianMaliszewski I propose to deal with `SSE` later on and for the moment focus on the web sockets only. So the web socket channel is initiated at the client's request and at this moment of time the bearer token must be available, the upgrade request is still coming on the plain HTTP as such this token is visible to the Quarkus code managing the upgrade to a web socket connection.
IMHO making this token a persistent property of the associated binding might work, so that the follow up web socket client requests would have this token automatically, but I'm not sure it is technically possible...

**Comment by DorianMaliszewski** [comment]
When I create my websocket through :
`const socket = new Websocket("ws://...")` I can't change headers during the connection upgrade. So where can I put this header during the connection upgrade ? 🤔 @sberyozkin

**Comment by sberyozkin** [comment]
Hi @DorianMaliszewski I believe there a way in general to set the attributes there but I've just read java script does not allow it. Is there some other way to pass the bearer token along but using the query params, @pedroigor does not recommend using them ?

**Comment by DorianMaliszewski** [comment]
I think i will, set an HTTP endpoint where the user will be authenticate, and give him a ticket stored in database to connect to the websocket. The user will use this ticket as parameter url to connect to the websocket. In the onOpen handler (`quarkus-undertow`) I get the ticket and authenticate the user.

@pedroigor @sberyozkin What do you think ? Do you understand what i'm planning ?

**Comment by sberyozkin** [comment]
@DorianMaliszewski I was reading about a ticket based mechanism [here](https://devcenter.heroku.com/articles/websocket-security#authentication-authorization), so yeah, it makes sense to me now :-). 
The bearer token will be effectively associated with the given web socket channel, I think it works though it is probably akin to creating a server-side session which is not very recommended for Quarkus (you'd need to manage the clients just closing the browsers for example, the DB may explode over time otherwise :-)).
And will this ticket travel to the server even after the upgrade, during every ws channel call ?

I thing the ticket idea is good, but it would be great if it could be just associated with a given web socket channel only, the client drops, the ticket has gone with the connection, @stuartwdouglas, can you advise something here ?

Thanks, Sergey

**Comment by DorianMaliszewski** [comment]
@sberyozkin 
I will store the ticket in a database, It will be nice in my case because I need all tracing, A ticket can be used only time and expire in 5 min after it is given.
I only check the ticket on onOpen method and after I store the session in a List.
To don't explode my DB, I will make some backup and truncate the table of old ticket

What do you think ?

**Comment by sberyozkin** [comment]
@DorianMaliszewski thanks, it makes sense. But if we can offer some OOB support for associating such custom tickets with the channel then it would simplify things a bit, the developers would only be concerned then with checking the actual ticket (which can be the original bearer token after all :-)) but I don't know much about the mechanics of how `ws:` is supported right now, lets wait for Stuart to comment.

**Comment by pedroigor** [comment]
That is an interesting discussion :)

I would say that the approach @DorianMaliszewski is proposing is pretty much based on the concept of a CSRF token. Where I think CSRF is the main type of attack you may want to protect your application against (e.g.: cookies as mentioned before, are vulnerable).

As such, by issuing this intermediary token you are basically giving to your clients a unique and random one-time-use token so that they can establish connections to your WS.

One thing I'm not sure though is how you are passing this token. By using a query parameter you are could potentially leak the token into server logs? I understand that the time window (e.g.: 5 minutes) is short but maybe enough time for someone to obtain and use it? Why not pass the token as if you were doing BASIC auth so that the token is sent as a header?

In regards to using OAuth2 access token to open WS connections, I think the main point is how to bound the token with a client so that the client can prove he is indeed in possession of that token (proof-of-possession). There are some specs in OAuth2 around that (as well as sender constraint tokens - authentication). Both bring complexity to a deployment :)

@sberyozkin I agree this issue is quite beyond OIDC and what we want to provide in `quarkus-oidc`.

**Comment by DorianMaliszewski** [comment]
@pedroigor 
Thank you to take part of this discussion. So IYO it's better to authenticate/authorize a user through basic authentication when initializing the ws ?

**Comment by pedroigor** [comment]
If it works, I would say is better than having the token as a query param.

**Comment by sberyozkin** [comment]
Hi Pedro, thanks for the feedback :-). Speaking of that, a smallrye-jwt user has opened a discussion about DPoP :-). Looks like we agree there should be a way of associating the ticket with the ws channel (so that it can really start acting as CSRF token, that was a helpful analogy)...

**Comment by pedroigor** [comment]
@sberyozkin Nice to hear. We should also start looking at that in Keycloak.

**Comment by gmussi** [comment]
I just came across this discussion as I am having the same conceptual problem (authenticated websocket).
So far, I am thinking of closing the stream if the first message in that stream is not a valid JWT. I would then manually validate the JWT.
I also liked the solution of the single-use ticket, but this would require additional domain logic, which I am trying to avoid.


What puzzled me is, in the Session of object, there is a .getUserPrincipal() method, which led me to believe there was a way to inject the user in that session somehow.

**Comment by DorianMaliszewski** [comment]
If it can help someone I made a flow with a one time ticket. I made one rest endpoint to generate a specific ticket and the websocket bean : I give you the classes.

```
@Authenticated
@Produces(MediaType.APPLICATION_JSON)
public class WebsocketRessource {

    @Inject
    SecurityIdentity securityIdentity;

    @Inject
    TicketService ticketService;

    @GET
    @Path("/get-ticket-for-room/{id}")
    public Response getTicket(@PathParam Long id) {
        String ticket = ticketService.createTicketForRoomId(id);
        Map<String, String> data = new HashMap<>();
        data.put("ticket", ticket);
        return Response.ok(data).build();
    }
}
```
Yes it's an ugly Map<String,String> but it do the job. Of course you can do an object :laughing: 

My model
```
@Entity
@Data
@RegisterForReflection
public class WebsocketTicketAccess extends PanacheEntityBase {
    private String username;

    @Id
    @GeneratedValue(generator = "UUID")
    @GenericGenerator(
            name = "UUID",
            strategy = "org.hibernate.id.UUIDGenerator"
    )
    @Column(name = "uuid", updatable = false, nullable = false)
    private String uuid;

    @CreationTimestamp
    private Date createdAt;

    @UpdateTimestamp
    private Date updatedAt;

    @Version
    public Integer version;

    private String tenant;
    private boolean used;

    @Column(name = "room_id")
    private Long roomId;
}
```


My Service
```
@RequestScoped
public class TicketService {
    private static final Logger LOG = Logger.getLogger(TicketService.class);


    @Inject
    SecurityIdentity identity;

    @Claim("tenant")
    String tenant;

    @Inject @RestClient
    UserService userService;

    @Inject
    ObjectMapper objectMapper;

    @Transactional
    public String createTicketForRoomId(Long roomId) {
        WebsocketTicketAccess websocketTicketAccess = new WebsocketTicketAccess();
        websocketTicketAccess.setTenant(tenant);
        websocketTicketAccess.setUsed(false);
        websocketTicketAccess.setUsername(identity.getPrincipal().getName());
        websocketTicketAccess.setRoomId(serviceId);
        websocketTicketAccess.persist();
        return websocketTicketAccess.getUuid().toString();
    }

    public WebsocketTicketAccess findById(String uuid) {
        return (WebsocketTicketAccess) WebsocketTicketAccess.findByIdOptional(uuid).orElseThrow(EntityNotFoundException::new);
    }

}
```

My websocket
```
@ApplicationScoped
@ServerEndpoint("/eztalk/ws/patient-requests/{tenant}/{ticket}")
public class PatientRequestWebSocket {

    @Inject
    WebsocketService websocketService;

    private final Executor executor = Executors.newSingleThreadExecutor();


    @OnOpen
    public void onOpen(Session session, @PathParam("ticket") String ticket) {
       // Obvious because it throw an error that you can block the thread
        CompletableFuture future = CompletableFuture.supplyAsync(() -> {
            websocketService.confirmAccess(ticket, session);
            return null;
        }, executor);
        future.join();
        System.out.println(ticket + " joined websocket");
    }

    @OnClose
    public void onClose(Session session, @PathParam("ticket") String ticket) {
        websocketService.closeSession(ticket);
        System.out.println(ticket + " left websocket");
    }

    @OnError
    public void onError(Session session, Throwable throwable, @PathParam("ticket") String ticket) {
        websocketService.closeSession(ticket);
        System.out.println(ticket + " left websocket on error: " + throwable);
    }

    @OnMessage
    public void onMessage(String message, @PathParam("ticket") String ticket) {
        // --
    }

    public void sendNewOrUpdate(AwesomeDTO awesomeDTO) {
        this.websocketService.sendUpdateToRoomId(awesomeDTO, awesomeDTO.getRoomId());
    }
}
```

My other service
```
@ApplicationScoped
@ActivateRequestContext
public class WebsocketService {

    private static final Logger LOG = Logger.getLogger(WebsocketService.class);


    @Inject
     ObjectMapper objectMapper;

    @Inject
    EntityManager entityManager;

    private Map<String, Session> sessions = new ConcurrentHashMap<>();

    @Transactional
    public javax.websocket.Session confirmAccess(String ticket, javax.websocket.Session session) {
        try {
            WebsocketTicketAccess ticketAccess = entityManager.createQuery("SELECT wta FROM WebsocketTicketAccess wta WHERE wta.uuid = :ticket", WebsocketTicketAccess.class)
                    .setParameter("ticket", ticket).getSingleResult();

            if (ticketAccess.isUsed()) {
                throw new ForbiddenException("This ticket is already used");
            }

            // I made all ticket good for 5 minutes
            if (ticketAccess.getCreatedAt().getTime() + (1000 * 60 * 5) < new Date().getTime()) {
                ticketAccess.setUsed(true);
                ticketAccess.persist();
                throw new ForbiddenException("This ticket has expired");
            }
            ticketAccess.setUsed(true);
            ticketAccess.persist();

            session.getUserProperties().put("tenant", ticketAccess.getTenant());
            session.getUserProperties().put("user", ticketAccess.getUsername());
            session.getUserProperties().put("roomId", ticketAccess.getRoomId());

            sessions.put(ticket, session);
            return session;
        } catch (ForbiddenException e) {
            try {
                session.close(new CloseReason(CloseReason.CloseCodes.UNEXPECTED_CONDITION, "Ticket not valid"));
            } catch (IOException ex) {
                ex.printStackTrace();
            }
            throw e;
        } catch (Exception e) {
            try {
                session.close(new CloseReason(CloseReason.CloseCodes.CLOSED_ABNORMALLY, "Server error"));
            } catch (IOException ex) {
                ex.printStackTrace();
            }
            throw e;
        }

    }


    public void sendUpdateToRoomId(AwesomeDTO awesomeDTO, Long roomId) {
        sessions.values().parallelStream().filter(session -> session.getUserProperties().get("roomId").equals(roomId))
                .forEach(session -> {
                    BaseMessage baseMessage = new BaseMessage();
                    baseMessage.setAction(WebsocketActionEnum.UPDATE.name());
                    try {
                        baseMessage.setContent(objectMapper.writeValueAsString(awesomeDTO));
                        session.getAsyncRemote().sendText(objectMapper.writeValueAsString(baseMessage), sendResult -> {
                            if (sendResult.getException() != null){
                                LOG.errorf("Unable to send message to %s, through websocket", session.getUserProperties().get("username"));
                            }
                        });
                    } catch (JsonProcessingException e) {
                        e.printStackTrace();
                        LOG.errorf("Unable to parse message to %s, through websocket", session.getUserProperties().get("username"));
                    }
                });
    }
    public void closeSession(String ticket){
        if (sessions.containsKey(ticket)) {
            try {
                sessions.get(ticket).close();
            } catch (IOException e) {
                LOG.error("Error when trying to close the session");
                e.printStackTrace();
            }
            sessions.remove(ticket);
        }
    }
}
```

I hope this will help you guys

**Comment by pedroigor** [comment]
@DorianMaliszewski Nice!

If we could just use a bounded hash map as well as make sure the client sending the ticket is indeed the one to which the token was first issued, it would be really nice.

**Comment by DorianMaliszewski** [comment]
We can add a field in WebsocketTicketAccess like "from" where we can store the subject of the token or the token completely and then made a findBy with UUID and from field ??

**Comment by sberyozkin** [comment]
@DorianMaliszewski Hi, interesting, how is the bearer token bound to this ticket ?

**Comment by DorianMaliszewski** [comment]
Sry forget to answer,

I used a claim in the token to bound it, you can store what you want, like the "jti", the "sub" which can be a god solution.

IMO if you use multiple OIDC server it's better to use the "jti" field. @pedroigor what do you think about that ?

**Comment by pedroigor** [comment]
It works to correlate both, but you are still not enforcing that only the client to which the ticket was issued for is indeed the same sending the ticket afterward. What I'm trying to solve is that token stolen by an attacker could be used to access your services. If when generating the ticket you add the certificate to your ticket you can check whether or not the client is the same when accessing your service.

There are specs from OAuth about how to solve this problem. They add complexity though, so I'm just highlighting here a possible vulnerability.

**Comment by sberyozkin** [comment]
Hi @DorianMaliszewski @pedroigor I'm trying to figure out if we should keep the issue open or there is some scope here for Quarkus web sockets or oidc to do something. I'd like to keep closing some OIDC issues :-).
We can add some utility support for binding the cert to the ticket, but I guess it can be easily done in the filter.
@DorianMaliszewski Are you OK if I close this issue ? May be you can blog about your solution ? thanks

**Comment by DorianMaliszewski** [comment]
It's ok no problem :)

**Comment by sberyozkin** [comment]
@DorianMaliszewski Thanks, please post the blog link once it is ready to the forum :-), I recall other users were asking about a similar issue

**Comment by ctron** [comment]
Is there a link to some blog post now? I have the same issue.
